src/Pjson.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types=1);
+
+namespace Square\Pjson;
+
+use ReflectionAttribute;
+use Square\Pjson\Internal\ArrayRecursionWrapper;
+use Square\Pjson\Internal\RClass;
+
+class Pjson
+{
+    public static function fromJsonString(
+        string $json,
+        string $type,
+        array|string $path = [],
+        int $depth = 512,
+        int $flags = 0,
+    ) {
+        $flags |= JSON_THROW_ON_ERROR;
+        $jd = json_decode($json, associative: true, flags: $flags, depth: $depth);
+        return static::fromJsonData($jd, $type, $path);
+    }
+
+    public static function listFromJsonString(
+        string $json,
+        string $type,
+        array|string $path = [],
+        int $depth = 512,
+        int $flags = 0,
+    ) : array {
+        $flags |= JSON_THROW_ON_ERROR;
+        $jd = json_decode($json, associative: true, flags: $flags, depth: $depth);
+        return static::listfromJsonData($jd, $type, $path);
+    }
+
+    public static function listfromJsonData(array $json, string $type, array|string $path = []) : array
+    {
+        if (is_string($path)) {
+            $path = [$path];
+        }
+        foreach ($path as $pathBit) {
+            $json = $json[$pathBit];
+        }
+        return array_map(fn ($d) => static::fromJsonData($d, $type), $json);
+    }
+
+    public static function fromJsonData($jd, string $type, array|string $path = [])
+    {
+        if (is_string($path)) {
+            $path = [$path];
+        }
+        foreach ($path as $pathBit) {
+            $jd = $jd[$pathBit];
+        }
+
+        $r = RClass::make($type);
+        if ($r->readsFromJson()) {
+            $continue = false;
+            if ($jd instanceof ArrayRecursionWrapper && $jd->type === $type) {
+                $continue = true;
+            }
+            if ($jd instanceof ArrayRecursionWrapper) {
+                $jd->type = $type;
+            }
+
+            if (!$continue) {
+                return $type::fromJsonData(ArrayRecursionWrapper::from($jd), $path);
```

Comment:
Alright, so if we're in the polymorphic use case: `CatalogObject` is a base class that implements `fromJsonData` to disambiguate between 2 child classes: `CatalogItem / CatalogCategory`.

Then those child classes also implement `fromJsonData` by virtue of extending the parent class. This means we'd get infinite recursion.

In the `trait` case, this is solved by having the child classes also `use JsonSerialize` then they don't call out to the parent anymore.
But here having a method like
```php
public function fromJsonData($data, $path) {
    return Pjson::fromJsonData($data, self::class);
}
```
Does not solve anything at all, it just sends us back to the same recursion.

So by wrapping the data in an object, we're able to figure out "have I already seen this exact data for this type". And if so, we bypass the `fromJsonData` class call and do our internal deserialization.

This works. 

Other solutions would be to:
- Force the polymorphic disambiguating class to not be a parent of the polymorphic classes. Though likely then the typehints start failing in the case of single properties so it'd have to be `public $object` with `#[Json(type: Disambiguator::class)]`
- require the use of the trait for this use case
- other solutions?

Reply:
I was trying hard to avoid adding a new interface / API surface for users of the library. `fromJsonData` is still required API surface if you look at the tests that do `BigInt` for example. And for doing more custom things so both `fromJsonData` and `discriminateJsonDataType` would become 2 possible ways of doing things in this suggestion.

Gonna let it simmer for a bit and see.